### ISSUE TITLE
SHAP with DQN

### ISSUE THREAD

**Issue by SimoMaestri** [shap/shap#2384]
Hi, i'm trying to apply shap.DeepExplainer to a DQN model. This is what i've done:
1. I've defined my model in this way:
`model = DQN('MlpPolicy', env,
                policy_kwargs=dict(net_arch=[256, 256]),
                learning_rate=5e-4,
                buffer_size=15000,
                learning_starts=200,
                batch_size=32,
                gamma=0.8,
                train_freq=1,
                gradient_steps=1,
                target_update_interval=50,
                exploration_fraction=0.7,
                verbose=1
                )`
2. Then i test my model and i've saved couples of state-action:
`def eval(video=False):
    #action_log = np.array([])
    action_log = []
    #state_log = np.array([])
    state_log = []
    
    
    env = gym.make("highway-fast-v0")
    if video==True:
        env = wrappers.Monitor(env, './videos/' + str(time()) + '/')
    s = env.reset()
    
    
    done = False
    reward = 0
    while not(done):
        if video:
            env.render()

        a, _ = model.predict(s, deterministic = True)   #ddpg.choose_action()
        s_, r, done, info = env.step(a)
        s = s_
        reward +=r
        

        action_log.append(a)
        state_log.append(s)
        
    return np.array(state_log), np.array(action_log)`
3. And then i use the function:
`state_log, action_log = eval()`
 
How can i apply shap.deepexplainer method? I mean, what input parameters should i give it in input?
I've tried to convert action-log and state-log to tensor and then pass them to shap.deepexplainer but this doesn't work.

**Comment by HarroTuin** [comment]
Hi, 

I'm facing a similar issue. I also use an algorithm (MaskablePPO) from the stable-baselines repo on a custom environment. I use the following:
1. Model: for the shap.DeepExplainer function I use ` model = model.policy.to(device) `
2. Data: I use the states that I saved during training (shape (45,)), this should be similar to your `state_log`

However, there seems to be an issue with normalizing these inputs. Eventually it throws this error:  
```
~\stable_baselines3\common\preprocessing.py in preprocess_obs(obs, observation_space, normalize_images)
    102         if is_image_space(observation_space) and normalize_images:
    103             return obs.float() / 255.0
--> 104         return obs.float()
    105 
    106     elif isinstance(observation_space, spaces.Discrete):
```

It seems something is going wrong with normalizing the input. I'll let you know if I figure it out.

Edit: I converted the input into Tensors, but now a different problems occurs.

**Comment by SimoMaestri** [comment]
Hi, i solved the problem in this way:
`explainer = shap.DeepExplainer(model.q_net, torch.FloatTensor(state_log).to(device))`
I used model = model.q_net that is the network between state and Q-value. Then i convert state_log to tensor (pytorch tensor because sb3 is based on pytorch).

**Comment by HarroTuin** [comment]
Thanks for your reply!
I tried something similar but I didn't get it to work. In the end I used[ this notebook](https://github.com/slundberg/shap/blob/master/notebooks/tabular_examples/neural_networks/Census%20income%20classification%20with%20Keras.ipynb) from the examples. This notebook uses the KernelExplainer instead of the DeepExplainer. For me this solved the problem.
